File: ios/Runner/MLAnalyzerPlugin.swift
import Flutter
import Foundation
import Combine

/// Native ML Analyzer Plugin for iOS
///
/// Acts as the View layer in MVVM architecture.
/// Binds MethodChannel inputs to ViewModel commands.
/// Binds ViewModel Published outputs to MethodChannel/EventChannel results.
class MLAnalyzerPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    
    static let channelName = "com.ai_assistant/ml_analyzer"
    static let eventChannelName = "com.ai_assistant/ml_events"
    
    private let viewModel = MLAnalyzerViewModel()
    private var cancellables = Set<AnyCancellable>()
    
    // For single-response methods
    private var pendingResult: FlutterResult?
    // For stream events
    private var eventSink: FlutterEventSink?
    
    static func register(with messenger: FlutterBinaryMessenger) {
        let instance = MLAnalyzerPlugin()
        
        // Register MethodChannel
        let channel = FlutterMethodChannel(name: channelName, binaryMessenger: messenger)
        channel.setMethodCallHandler(instance.handle)
        
        // Register EventChannel
        let eventChannel = FlutterEventChannel(name: eventChannelName, binaryMessenger: messenger)
        eventChannel.setStreamHandler(instance)
        
        instance.observeViewModel()
    }
    
    static func register(with registrar: FlutterPluginRegistrar) {
        // Compatibility helper if needed, but main entry point is typically register(with messenger)
        // for modern Flutter iOS plugins or AppDelegate registration.
        // We reuse the logic by extracting the messenger.
        register(with: registrar.messenger())
    }
    
    private func observeViewModel() {
        viewModel.$analysisState
            .receive(on: RunLoop.main)
            .compactMap { $0 } // Filter out initial nil
            .sink { [weak self] state in
                // Send to pending result (Request-Response pattern)
                if let result = self?.pendingResult {
                    result(state)
                    self?.pendingResult = nil
                }
                
                // Also broadcast to EventChannel (Stream pattern)
                if let sink = self?.eventSink {
                    sink(state)
                }
            }
            .store(in: &cancellables)
    }
    
    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "analyzeMessage":
            guard let args = call.arguments as? [String: Any],
                  let message = args["message"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Message is required", details: nil))
                return
            }
            pendingResult = result
            viewModel.analyze(message)
            
        case "isAvailable":
            result(true)
            
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    // --- FlutterStreamHandler Implementation ---
    
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
}

File: android/app/src/main/kotlin/com/example/ai_assistant/MLAnalyzerPlugin.kt
package com.example.ai_assistant

import android.content.Context
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

/**
 * Native ML Analyzer Plugin for Android
 * 
 * Acts as the View layer in MVVM architecture.
 * Binds MethodChannel inputs to ViewModel commands.
 * Binds ViewModel StateFlow outputs to MethodChannel/EventChannel results.
 */
class MLAnalyzerPlugin private constructor(
    private val context: Context
) : MethodChannel.MethodCallHandler, EventChannel.StreamHandler {

    private val viewModel = MLAnalyzerViewModel(context)
    private val mainScope = CoroutineScope(Dispatchers.Main)
    
    // For single-response methods
    private var pendingResult: MethodChannel.Result? = null

    companion object {
        private const val CHANNEL_NAME = "com.ai_assistant/ml_analyzer"
        private const val EVENT_CHANNEL_NAME = "com.ai_assistant/ml_events"

        fun registerWith(flutterEngine: FlutterEngine, context: Context) {
            val plugin = MLAnalyzerPlugin(context)
            
            // Register MethodChannel
            val methodChannel = MethodChannel(flutterEngine.dartExecutor.binaryMessenger, CHANNEL_NAME)
            methodChannel.setMethodCallHandler(plugin)
            
            // Register EventChannel
            val eventChannel = EventChannel(flutterEngine.dartExecutor.binaryMessenger, EVENT_CHANNEL_NAME)
            eventChannel.setStreamHandler(plugin)
            
            plugin.observeViewModel()
        }
    }

    private fun observeViewModel() {
        mainScope.launch {
            viewModel.analysisState.collect { state ->
                if (state != null) {
                    // Send to pending result (Request-Response pattern)
                    pendingResult?.success(state)
                    pendingResult = null
                    
                    // Also broadcast to EventChannel (Stream pattern)
                    eventSink?.success(state)
                }
            }
        }
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "analyzeMessage" -> {
                val message = call.argument<String>("message")
                if (message != null) {
                    pendingResult = result
                    viewModel.analyze(message)
                } else {
                    result.error("INVALID_ARGUMENT", "Message is required", null)
                }
            }
            "isAvailable" -> {
                // Now checking the real availability of the engine
                result.success(viewModel.isAvailable())
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    // --- EventChannel Implementation ---
    private var eventSink: EventChannel.EventSink? = null

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }
    
    fun dispose() {
        mainScope.cancel()
    }
}
